buildSrc/src/main/kotlin/net/skyscanner/backpack/tokens/BpkTextStyles.kt

```diff
@@ -0,0 +1,180 @@
+
+    override fun invoke(source: Map<String, Any>): BpkTextStyles =
+      parseTextStyles(source)
+  }
+
+  sealed class Format : BpkTransformer<BpkTextStyles, TypeSpec> {
+
+    data class Compose(val className: String) : BpkTextStyle.Format() {
+      override fun invoke(source: BpkTextStyles): TypeSpec =
+        toCompose(source, className)
+    }
+  }
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun parseTextStyles(source: Map<String, Any>): BpkTextStyles {
+
+  fun String.trimReference(): String =
+    removePrefix("{!").removeSuffix("}")
+
+  fun String.toSemanticName() =
+    CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, this.replace('-', '_'))
+
+  fun String.wrapIfDigits() = if (first().isDigit()) { "`$this`" } else this
+
+  fun List<Map.Entry<String, Map<String, String>>>.findTokenName(type: String): String? {
+    return firstOrNull { it.key.endsWith(type) }
+      ?.let {
+        it.value["originalValue"]!!
+          .trimReference()
+          .removePrefix("${type}_")
+          .toSemanticName()
+          .wrapIfDigits()
+      }
+  }
+
+  val props = source.getValue("props") as Map<String, Map<String, String>>
+  val data = props.filter { (key, value) ->
+    value["deprecated"] != "true" &&
+      key.startsWith("TEXT_") &&
+      value["type"] != "color"
+  }
+
+  val list = data
+    .mapKeys { it.key.removePrefix("TEXT_") }
+    .entries.groupBy { it.key.substringBefore("_") }
+    .map {
+      BpkTextStyleModel(
+        name = it.key,
+        fontSize = it.value.findTokenName("FONT_SIZE")!!,
+        fontWeight = it.value.findTokenName("FONT_WEIGHT")!!.let { if (it == "Book") "Normal" else it },
+        lineHeight = it.value.findTokenName("LINE_HEIGHT")!!,
+        letterSpacing = it.value.findTokenName("LETTER_SPACING"),
+      )
+    }
+
+  return object : BpkTextStyles, List<BpkTextStyleModel> by list {
+    override fun toString(): String =
+      list.toString()
+  }
+}
+
+private val TextStyleClass = ClassName("androidx.compose.ui.text", "TextStyle")
+private val FontFamilyClass = ClassName("androidx.compose.ui.text.font", "FontFamily")
+private val FontWeightClass = ClassName("androidx.compose.ui.text.font", "FontWeight")
+private val BpkFontSizeClass = ClassName("net.skyscanner.backpack.compose.tokens", "BpkFontSize")
+private val BpkLineHeightClass = ClassName("net.skyscanner.backpack.compose.tokens", "BpkLineHeight")
+private val BpkLetterSpacingClass = ClassName("net.skyscanner.backpack.compose.tokens", "BpkLetterSpacing")
+
+
+private fun toCompose(source: BpkTextStyles, className: String): TypeSpec {
+  val defaultFontFamily = "defaultFontFamily"
+
+  fun String.toSemanticName() =
+    CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, this.replace('-', '_'))
+
+  fun BpkTextStyles.toTextStyleClass(): TypeSpec {
+
+    fun BpkTextStyles.toConstructorFunction(): FunSpec {
+
+      fun BpkTextStyleModel.toParameter(): String =
+        "%textStyle:T(\n" +
```

Comment:
unfortunately there's no nice helper methods for the constructor function itself, so we have to build this all ourselves. the `%x:T` syntax is for named placeholders, so we can refer to class names etc